### ISSUE TITLE
Strict types for sqlite

### ISSUE THREAD

**Issue by NorfairKing** [yesodweb/persistent#1542]
Sqlite got support for strict tables: https://www.sqlite.org/stricttables.html

Could they be enabled by default when using persistent?

**Comment by malteneuss** [comment]
I would also like to have this feature supported and enabled. There would probably need to be some changes to the migration sql code generation, because the generated sql types would need to match the few supported sqlite types then; without STRICT you apparently can mention any type and sqlite will just accept it, which what persistent is doing as of now.

**Comment by malteneuss** [comment]
I'm not sure how this is related, but might be an additional issue.

What i mean is that if you have e.g.

```haskell
share
    [mkPersist sqlSettings , mkMigrate "migrateAll"]
    [persistLowerCase|
        SomeEntity 
            Id UUID
            url Text
            createdAt UTCTime
        |]
```

then the generated migration code is

```sql
CREATE TABLE "some_entity"(
    "id" VARCHAR PRIMARY KEY,
    "url" VARCHAR NOT NULL,
    "created_at" TIMESTAMP NOT NULL
);
```

and works, because sqlite (non-strict) accepts anything. However, with strict, i.e.

```sql
CREATE TABLE "some_entity"(
    "id" VARCHAR PRIMARY KEY,
    "url" VARCHAR NOT NULL,
    "created_at" TIMESTAMP NOT NULL
) STRICT;
```

those types aren't valid sqlite types so we get e.g. `Error: unknown datatype for some_entity.id: "VARCHAR"`. 

So the sqlite backend type mappings need to change to one of following:

    INT
    INTEGER
    REAL
    TEXT
    BLOB
    ANY 

is what i mean.

**Comment by NorfairKing** [comment]
Ah yes that's a great point!

**Comment by malteneuss** [comment]
Regarding more strictness, i added a small "bugfix" in #1585. @parsonsmatt could you or someone other maintainer have a look. Afterwards i could try to tackle the STRICT migration, but will probably need some more permissions to read the CI logs for errors (using NixOS and `nix develop` i can't run many integration tests locally so far)

**Comment by malteneuss** [comment]
I started a first draft in https://github.com/yesodweb/persistent/pull/1590 (not tested yet) although i'm not sure this is really backwards compatible or correct. 

Correct me if i'm wrong, but Persistent funnels user data types into a generic set of SqlTypes/PersistValue (where you have to support all cases even though there's no clear mapping to column types for some) and maps those to the specific database column representation. In many databases, especially Sqlite, there are many column encodings that the user should be able to choose. E.g. UTCDateTime can either be stored as in UNIX timestamp integer or as a ISO8601 text string. A likely better approach for a 3.0 version would be to have a SqlTypes/PersisteValue type for every backend like Rust's sqlx does: https://github.com/launchbadge/sqlx/blob/main/sqlx-sqlite/src/type_info.rs#L13

**Comment by parsonsmatt** [comment]
Do you mean like a runtime configurable function like `backendSqlType :: SqlBackend -> SqlType -> Text`? Generally backend libraries are responsible for converting `SqlType` into something the backend can use reliably. 

You could add a record field to `SqlBackend` with, say, `customSqlType :: SqlType -> Maybe Text` that would allow the given backend to override the way that the type is used. Then the backend library itself would be modified to use `fromMaybe (defaultSqlType sqlType) (customSqlType sqlType)`. However, there's no guarantee that *other* uses of `SqlType` would be converted according to the `SqlBackend`. If we want to require that, then we'd need for `sqlType :: SqlBackend -> Proxy a -> SqlType`, which is a significant breaking change.

**Comment by malteneuss** [comment]
No, not at runtime, but at compile time. Now that i had more to think, my source of confusion is a design quirk in Persistent wrt. to Sqlite where you need to have **two mappings per type in mind** (an additional layer of indirection).

Let's look at a specific example:

```
share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
Person
    isAdult Bool <----
|]
```

A value of **type `Bool` is mapped to an intermediate representation `PersistBool :: PersistValue`(used for all backends), which is [actually mapped to an integer](https://github.com/yesodweb/persistent/blob/master/persistent-sqlite/Database/Sqlite.hs#L468)** in the Sqlite database.
A **direct mapping `Bool` to a hypothetical `PersistInteger :: PersistSqliteValue``** (that doesn't have PersistBool, because **Bool doesn't exist in Sqlite**) like in Sqlx in Rust would be mentally easier. But i guess that's would be too much churn for now. Maybe in a Persistent 3.0.

**Comment by malteneuss** [comment]
I now think that we cannot introduce STRICT without a breaking change. 

For example the `Rational` type is actually written to Sqlite as text https://github.com/yesodweb/persistent/blob/594a5322819bc15100abeb379bdf318d8f89fa1f/persistent-sqlite/Database/Sqlite.hs#L467. And because the current Sqlite ["type" is NUMERIC](https://github.com/yesodweb/persistent/blob/594a5322819bc15100abeb379bdf318d8f89fa1f/persistent-sqlite/Database/Persist/Sqlite.hs#L440) (not a real type but a class of implicit conversions), a rational could have been stored as a float or an integer. If i now use a strict Sqlite type i have to choose either "INTEGER" or "REAL", which **i think might break existing databases** (correct me if i'm wrong).. So, we may want to use STRICT only in a Persistent 3.0.

**Comment by parsonsmatt** [comment]
Another alternative - albeit a bit gross - is to create a fork of `persistent-sqlite` named `persistent-sqlite-strict` that has a different interpretation of `PersistValue` in accordance with strict mode.

**Comment by NorfairKing** [comment]
@parsonsmatt I'd be in favour of that because it won't cause breakage with people who don't want strict types enable, but it won't prevent breakage for people who want to switch and still read old databases.
